### PR TITLE
Mitigate out of free space issues while running dartdoc postprocessing.

### PR DESCRIPTION
**Pull request by isoos**
- #7981
- By deleting the input dartdoc data, the post-processing in both isolates will stop.
- We are deleting the partial output as we won't use it, and we may need free space to complete writing the log.txt and the blob files.

### CODE REVIEW

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart [hunk 1]
```diff
@@ -85,14 +85,39 @@ Future<void> postProcessDartdoc({
...
+    // This helps if there is a timeout along the way.
+    await tmpOutDir.rename(p.join(outputFolder, 'doc'));
+    await tmpTar.rename(p.join(outputFolder, 'doc', 'package.tar.gz'));
+  } catch (e, st) {
+    // We are likely out of free disk space.
+    //
+    // One of the isolate has already stopped, the other may be still working.
+    // By deleting the dartdoc-generated content, the post-processing of the other
+    // isolate will stop too.
+    await Directory(docDir).deleteIgnoringErrors(recursive: true);
```
Comment (jonasfj) [inline pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:105]: I don't understand this.

Who owns the `docDir`, I think we should say that it's owned by the caller to `postProcessDartdoc` as is currently the case.

And that `postProcessDartdoc` only read from `docDir`.

It feels messy to be sometimes deleting it here and sometimes not.
If we want to delete it here, we should always do so, and we should document it for the caller.

However, I see no reason to do that.

--------------------------

Here we're trying to stop the isolates from running by deleting their input files.
I think that's a bit wild. Perhaps it'd be more reasonable to:
 * (A) Not run the two thing concurrently, so we don't have to use isolates.
 * (B) Create a `runInIsolate<T>(FutureOr<T> Function() computation, {Completer<void> cancel})`.

Obviously, (B) would require that we create a tiny helper library for running stuff in an isolate with cancellation. And that we have tests :rofl: 

I think the correct approach for what you're trying to do is (B), but that the simply solution is (A). Afterall, it's fairly likely that both these operations are going to be I/O bound (maybe not post-processing).

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart [hunk 2]
```diff
@@ -85,14 +85,39 @@ Future<void> postProcessDartdoc({
...
-  await tmpTar.rename(p.join(outputFolder, 'doc', 'package.tar.gz'));
+    // Move from temporary output directory to final one, ensuring that
+    // documentation files won't be present unless all files have been processed.
+    // This helps if there is a timeout along the way.
+    await tmpOutDir.rename(p.join(outputFolder, 'doc'));
+    await tmpTar.rename(p.join(outputFolder, 'doc', 'package.tar.gz'));
+  } catch (e, st) {
+    // We are likely out of free disk space.
+    //
+    // One of the isolate has already stopped, the other may be still working.
```
Comment (jonasfj) [inline pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:102]: That's not the semantics of [`Future.wait`](https://api.dart.dev/stable/3.5.3/dart-async/Future/wait.html).

File: pkg/pub_worker/lib/src/bin/pana_wrapper.dart [hunk 3]
```diff
@@ -167,7 +167,10 @@ Future<void> main(List<String> args) async {
     docDir: rawDartdocOutputFolder.path,
   );
 
-  await rawDartdocOutputFolder.delete(recursive: true);
+  // existence check is needed as post-processing may delete this directory
```
Comment (jonasfj) [inline pkg/pub_worker/lib/src/bin/pana_wrapper.dart:170]: When this method creates the folder, this method should probably also delete it.

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart (hunk 2 above)
Comment (isoos) [reply pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart]: Good catch. Fixed it. We may wait for both, as it is likely that both will be running out of the free space the same time.

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart (hunk 1 above)
Comment (isoos) [reply pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart]: Fixed the `docDir` deletion. Agreed on (B) being the better one, but for a simple handling we may just do the cleanup.

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart [hunk 4]
```diff
@@ -85,14 +85,33 @@ Future<void> postProcessDartdoc({
...
+    await tmpOutDir.rename(p.join(outputFolder, 'doc'));
+    await tmpTar.rename(p.join(outputFolder, 'doc', 'package.tar.gz'));
+  } catch (e, st) {
+    // We are likely out of free disk space.
+    // Cleanup the content as we won't use it.
+    await tmpOutDir.deleteIgnoringErrors(recursive: true);
+    await tmpTar.deleteIgnoringErrors();
+
+    _log.shout('Failed to run dartdoc post-processing.', e, st);
+  }
```
Comment (jonasfj) [inline pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:106]: So now now it'll just complete successfully without any content.

Are we sure that's the behavior we want?

The log messages here go to stackdriver. Not the package task-log.

Which is IMO also the right thing.

But successfully completing the task is probably not desirable. Or at-least if we do, we should write in the task-log that something failed.

Comment (jonasfj) [reply pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:106]: To be fair, I'm not sure we fully understand what it is we want to do. We should perhaps chat a bit more about this.

File: pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart [hunk 5]
```diff
@@ -85,14 +85,33 @@ Future<void> postProcessDartdoc({
...
+    // This helps if there is a timeout along the way.
+    await tmpOutDir.rename(p.join(outputFolder, 'doc'));
+    await tmpTar.rename(p.join(outputFolder, 'doc', 'package.tar.gz'));
+  } catch (e, st) {
+    // We are likely out of free disk space.
+    // Cleanup the content as we won't use it.
+    await tmpOutDir.deleteIgnoringErrors(recursive: true);
+    await tmpTar.deleteIgnoringErrors();
+
+    _log.shout('Failed to run dartdoc post-processing.', e, st);
```
Comment (jonasfj) [inline pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:105]: At a minimum we should exit non-zero here, right?

Comment (jonasfj) [reply pkg/pub_worker/lib/src/bin/dartdoc_wrapper.dart:105]: or rethrow?

**Review by jonasfj — APPROVED** [review]

**Pull request merged**